File: be_camphub/src/main/java/org/camphub/be_camphub/mapper/ReviewMapper.java
package org.camphub.be_camphub.mapper;

import org.camphub.be_camphub.dto.request.review.ReviewCreationRequest;
import org.camphub.be_camphub.dto.response.review.ReviewResponse;
import org.camphub.be_camphub.entity.Review;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public interface ReviewMapper {
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())")
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "comment", source = "content")
    Review toEntity(ReviewCreationRequest request);

    @Mapping(target = "reviewerName", ignore = true) // sẽ set trong service
    @Mapping(target = "itemName", ignore = true) // sẽ set trong service
    @Mapping(target = "reviewerAvatar", ignore = true) // sẽ set trong service
    ReviewResponse toResponse(Review entity);
}

File: fe_camphub/src/libs/core/mapping/review.mapper.ts
import { createMap, createMapper, forMember, mapFrom } from "@automapper/core";
import { pojos, PojosMetadataMap } from "@automapper/pojos";
import { Review } from "../types";
import { ReviewResponse } from "../dto/response";
import { ReviewCreationRequest } from "../dto/request";

// ------------------------
// Mapper
// ------------------------
export const reviewMapper = createMapper({
    strategyInitializer: pojos(),
});

// ------------------------
// Metadata
// ------------------------
PojosMetadataMap.create<Review>("Review", {
    id: String,
    bookingId: String,
    reviewerId: String,
    reviewedId: String,
    itemName: String,
    reviewerName: String,
    reviewerAvatar: String,
    rating: Number,
    comment: String,
    mediaUrls: Array,
    createdAt: String,
});
PojosMetadataMap.create<ReviewResponse>("ReviewResponse", {
    id: String,
    bookingId: String,
    reviewerId: String,
    reviewedId: String,
    itemName: String,
    reviewerName: String,
    reviewerAvatar: String,
    rating: Number,
    comment: String,
    mediaUrls: Array,
    createdAt: String,
});
PojosMetadataMap.create<ReviewCreationRequest>("ReviewCreationRequest", {
    bookingId: String,
    reviewerId: String,
    reviewedId: String,
    rating: Number,
    content: String,
    mediaUrls: Array,
});

// ------------------------
// Mapping: DTO response → Model
// ------------------------
createMap<ReviewResponse, Review>(
    reviewMapper,
    "ReviewResponse",
    "Review"
);

// ------------------------
// Mapping: Model → Creation Request DTO
// ------------------------
createMap<Review, ReviewCreationRequest>(
    reviewMapper,
    "Review",
    "ReviewCreationRequest",
    // map content từ comment
    forMember(d => d.content, mapFrom(s => s.comment))
);

// ------------------------
// Export API
// ------------------------
export const reviewMap = {
    fromResponse: (response: ReviewResponse): Review =>
        reviewMapper.map<ReviewResponse, Review>(
            response,
            "ReviewResponse",
            "Review"
        ),
    toCreationRequest: (model: Review): ReviewCreationRequest =>
        reviewMapper.map<Review, ReviewCreationRequest>(
            model,
            "Review",
            "ReviewCreationRequest"
        ),
};
